From 099ef7e9bceae1eddaeedb7ff6c6ef95ced02935 Mon Sep 17 00:00:00 2001
From: Tim-Philipp Muller <tim@centricular.com>
Date: Mon, 26 Dec 2022 11:56:42 +0000
Subject: [PATCH 05/11] gtypes.h: workaround for g_uint64_checked_mul() issue
 with clang on 32-bit ARM

gdk-pixbuf would fail to build on Android on armv7, with:

	error: undefined reference to '__multi3'

and on x86 with

	error: undefined reference to '__mulodi4'

which looks like a toolchain issue, so hack around it for now by making
it not use the builtins in that case.

https://reviews.llvm.org/D103906
https://bugs.llvm.org/show_bug.cgi?id=16404
https://gitlab.freedesktop.org/gstreamer/cerbero/-/merge_requests/864#note_1698211
---
 glib/gtypes.h | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/glib/gtypes.h b/glib/gtypes.h
index 9d912d523..dc923b60e 100644
--- a/glib/gtypes.h
+++ b/glib/gtypes.h
@@ -456,6 +456,14 @@ typedef const gchar *   (*GTranslateFunc)       (const gchar   *str,
 #endif
 #endif
 
+/* HACK: work around clang toolchain issue on android armv7/x86 compiling gdk-pixbuf
+ * in GStreamer cerbero: https://reviews.llvm.org/D103906 (undefined ref to __multi3)
+ * Android (5900059 based on r365631c) clang version 9.0.8
+ * https://gitlab.freedesktop.org/gstreamer/cerbero/-/merge_requests/864#note_1698211 */
+#if defined(__clang__) && !defined(CRT_HAS_128BIT)
+#undef _GLIB_HAVE_BUILTIN_OVERFLOW_CHECKS
+#endif
+
 #ifdef _GLIB_HAVE_BUILTIN_OVERFLOW_CHECKS
 
 #define g_uint_checked_add(dest, a, b) \
-- 
2.44.0.windows.1

